### PR TITLE
ACS-8464 user agent header in hxinsight JAMP

### PR DESCRIPTION
**Pull request by mpichura**
Also abstracting ACS version retrieval.

### CODE REVIEW

File: hxinsight-extension/src/main/resources/alfresco/module/alfresco-hxinsight-connector-hxinsight-extension/context/service-context.xml [hunk 1]
```diff
@@ -174,4 +174,15 @@
     <bean class="org.alfresco.hxi_connector.hxi_extension.rest.api.QuestionFeedbackRelation">
         <constructor-arg name="hxInsightClient" ref="hxInsightClient"/>
     </bean>
+
+    <bean id="acsRepositoryInformation" class="org.alfresco.hxi_connector.hxi_extension.service.AcsRepositoryInformation">
+        <constructor-arg name="descriptorService" ref="DescriptorService"/>
+    </bean>
+    <bean id="hxiApplication" class="org.alfresco.hxi_connector.common.config.properties.Application">
```
Comment (krdabrowski) [inline hxinsight-extension/src/main/resources/alfresco/module/alfresco-hxinsight-connector-hxinsight-extension/context/service-context.xml:181]: The bean name (and also property bellow) is confusing for me - I assume it's connector version, not some HxI AI app version

**Review by krdabrowski — APPROVED** [review]
LGTM, I have only one remark

**Review by damianujma — APPROVED** [review]
LGTM.

File: hxinsight-extension/src/main/resources/alfresco/module/alfresco-hxinsight-connector-hxinsight-extension/context/service-context.xml (hunk 1 above)
Comment (mpichura) [reply hxinsight-extension/src/main/resources/alfresco/module/alfresco-hxinsight-connector-hxinsight-extension/context/service-context.xml]: Changed - hopefully new names are better suiting and less confusing.

**Review by aonych — APPROVED** [review]

File: hxinsight-extension/src/main/java/org/alfresco/hxi_connector/hxi_extension/service/HxInsightClient.java [hunk 2]
```diff
@@ -104,7 +109,8 @@ public String askQuestion(Question question)
 
             HttpRequest request = HttpRequest.newBuilder()
                     .uri(URI.create(config.getQuestionUrl()))
-                    .header("Content-Type", "application/json")
+                    .header(CONTENT_TYPE, "application/json")
+                    .header(USER_AGENT, applicationInfoProvider.getUserAgentData())
```
Comment (dsibilio) [inline hxinsight-extension/src/main/java/org/alfresco/hxi_connector/hxi_extension/service/HxInsightClient.java:113]: I wonder if we could have solved this with a solution that is more fit to address cross-cutting concerns such as this one, e.g.: using an HttpRequestInterceptor?

I'm personally ok with this solution, but it is error-prone in the sense that all new APIs that are implemented rely on the developers to remember they have to manually add this user agent calculation into the request.

File: live-ingester/src/main/java/org/alfresco/hxi_connector/live_ingester/adapters/config/messaging/LiveIngesterMessagingConfig.java [hunk 3]
```diff
@@ -62,20 +63,24 @@ public Application application()
...
+        else if (StringUtils.isNotBlank(integrationProperties.alfresco().repository().discoveryEndpoint()))
+        {
+            return new DiscoveryApiRepositoryInformation(integrationProperties.alfresco().repository().discoveryEndpoint(), authService, objectMapper, HttpClient.newHttpClient());
+        }
+        throw new IllegalStateException("Either property alfresco.repository.discovery-endpoint or alfresco.repository.version-override must be set in the Live Ingester configuration.");
     }
 
     @Bean
-    public ApplicationInfoProvider applicationInfoProvider(DiscoveryApiClient discoveryApiClient, IntegrationProperties integrationProperties)
+    @DependsOn("repositoryInformation")
```
Comment (dsibilio) [inline live-ingester/src/main/java/org/alfresco/hxi_connector/live_ingester/adapters/config/messaging/LiveIngesterMessagingConfig.java:80]: I find it odd that this `DependsOn` was necessary even though `RepositoryInformation` is one of the required arguments (read: beans) of the method. Could this be an intermediate change that is not required anymore?

File: common-authentication/src/main/java/org/alfresco/hxi_connector/common/adapters/messaging/repository/DiscoveryApiRepositoryInformation.java [hunk 4]
```diff
@@ -81,7 +85,7 @@ private DiscoveryApiResponse getDiscoverApiResponse() throws IOException, Interr
         throwExceptionOnUnexpectedStatusCode(response.statusCode(), SC_OK);
 
         DiscoveryApiResponse discoveryApiResponse = objectMapper.readValue(response.body(), DiscoveryApiResponse.class);
-        log.atTrace().log("Discovery API response: {}", discoveryApiResponse);
+        log.trace("Discovery API response: {}", discoveryApiResponse);
```
Comment (dsibilio) [inline common-authentication/src/main/java/org/alfresco/hxi_connector/common/adapters/messaging/repository/DiscoveryApiRepositoryInformation.java:88]: I meant to ask, is this API only invoked once at the startup to determine the version, or is it invoked N times? If it's the latter, then we should - at least eventually - plan caching this IMO.

**Review by dsibilio — COMMENTED** [review]
Mostly LGTM, left a few comments with doubts/requests for clarifications and suggestions for potential improvements. I'll leave it up to you to decide whether it's appropriate to address those now or it would be best to track them and deal with them later on.

File: hxinsight-extension/src/main/java/org/alfresco/hxi_connector/hxi_extension/service/HxInsightClient.java (hunk 2 above)
Comment (mpichura) [reply hxinsight-extension/src/main/java/org/alfresco/hxi_connector/hxi_extension/service/HxInsightClient.java:113]: I tried to use a more reliable solution but with `java.net.http.HttpClient` there's not much to do here.
I guess the best thing we could possibly do is some custom implementation (either for `java.net.http.HttpRequest` or maybe `java.net.http.HttpClient`) which would cover that. 
I'll push some more generic approach for this - let me know what you think.

File: common-authentication/src/main/java/org/alfresco/hxi_connector/common/adapters/messaging/repository/DiscoveryApiRepositoryInformation.java (hunk 4 above)
Comment (mpichura) [reply common-authentication/src/main/java/org/alfresco/hxi_connector/common/adapters/messaging/repository/DiscoveryApiRepositoryInformation.java:88]: Thanks for catching the logger part.
About caching - `ApplicationInfoProvider` class which is used across all modules as `User-Agent` header value provider will call `RepositoryInformation::getRepositoryVersion` only once and keep it in class member field.
If this solution is somehow faulty, we can plan some caching, but for now, I don't see a reason to do so.

File: live-ingester/src/main/java/org/alfresco/hxi_connector/live_ingester/adapters/config/messaging/LiveIngesterMessagingConfig.java (hunk 3 above)
Comment (mpichura) [reply live-ingester/src/main/java/org/alfresco/hxi_connector/live_ingester/adapters/config/messaging/LiveIngesterMessagingConfig.java:80]: Thanks for catching that - probably a leftover from some try-and-test - removing that annotation.

File: hxinsight-extension/src/main/java/org/alfresco/hxi_connector/hxi_extension/service/HxInsightClient.java (hunk 2 above)
Comment (dsibilio) [reply hxinsight-extension/src/main/java/org/alfresco/hxi_connector/hxi_extension/service/HxInsightClient.java]: We could extend the `HttpClient#send` method to automatically decorate all requests, creating an e.g.: `HxHttpClient` class.

Similar to https://stackoverflow.com/questions/53215038/how-to-log-request-response-using-java-net-http-httpclient/55455589#55455589.

**Pull request merged**